### ISSUE TITLE
How to add multiple edges in single query faster?

### ISSUE THREAD

**Issue by Munmud** [apache/age#918]
Suppose I have a graph with maybe 3k nodes. Now I have to add a lot of edges at a time. Adding edges one by one will take a lot of time due to it needs to search among all nodes and add edges. So is there any faster way to add multiple edges at a time. Maybe in a single cypher query ?

**Comment by farooquememon385** [comment]
**You can add edges by running query with a rule between to node Types**
### for example:

we have Person nodes and Country nodes in graph named 'GRAPH_NAME' and we want to make a relation that Person born in some country. 

**we can make relationship by running the query like this: **

SELECT * FROM cypher('GRAPH_NAME', $$ MATCH (a:Person), (b:Country) WHERE a.bornIn = b.name CREATE (a)-[r:BORNIN]->(b) RETURN r $$) as (r agtype);

no it does not matter that we have how many nodes of Person or Country
it will just create relationship based on the rule that **Person.bornIn = Country.name**

**Comment by titoausten** [comment]
Firstly, you should have a basis for creating edges and should have an idea of the nodes in the graph.

For instance, if you have nodes with labels `Employee`, `Company` and `Role` and you want to add edges between them. Each node should have a property that might be similar to a property in another node.

Now to create edges between nodes with labels `Employee` and `Company`, you simply use the `MATCH` clause. This will look like:
```
SELECT * FROM cypher('graph', $$ MATCH (a:Employee), (b:Company) WHERE a.company = b.name CREATE (a)-[r:worksAt]->(b) RETURN r $$) as (r agtype);
```
To create edges between nodes with labels `Employee` and `Role`, the query will be:
```
SELECT * FROM cypher('graph', $$ MATCH (a:Employee), (b:Role) WHERE a.position = b.name CREATE (a)-[r:IS]->(b) RETURN r $$) as (r agtype);
```
In conclusion, creating multiple edges is dependent on matching properties of nodes.

**Comment by humzakt** [comment]
In Apache AGE, you can indeed insert multiple edges in a single Cypher query which can help optimize the operation. The Cypher query language used in AGE allows the creation of multiple relationships (edges) in a single command. Below is an example of how you can do this:

Assuming you have two sets of nodes:
```
CREATE (a:Person {name: 'John'}), (b:Person {name: 'Jack'}), (c:Person {name: 'Jill'})
CREATE (x:Location {name: 'New York'}), (y:Location {name: 'Los Angeles'}), (z:Location {name: 'Chicago'})
```

You can create multiple edges in a single command as follows:

```
MATCH (a:Person {name: 'John'}), (b:Person {name: 'Jack'}), (c:Person {name: 'Jill'})
MATCH (x:Location {name: 'New York'}), (y:Location {name: 'Los Angeles'}), (z:Location {name: 'Chicago'})
CREATE (a)-[:VISITED]->(x), (b)-[:VISITED]->(y), (c)-[:VISITED]->(z)
```

In the above example, three edges are created in a single Cypher command. This is more efficient than creating them one by one. Make sure to use indexes on properties that are used often in MATCH clauses to further speed up the process.

However, please note that if you're trying to add a large number of edges at once, you might hit transaction size limits depending on your PostgreSQL configuration. In such cases, you may need to batch the operations or adjust your PostgreSQL configuration.

**Comment by Amr-Shams** [comment]
when it comes to this large size it is better to make it automatically by creating a function that takes the list and iterates over the edges and specified information and the function will do the job for your but still need to process the list before passing to the function.. like creating a list of objects...

**Comment by Munmud** [comment]
> when it comes to this large size it is better to make it automatically by creating a function that takes the list and iterates over the edges and specified information and the function will do the job for your but still need to process the list before passing to the function.. like creating a list of objects...

@Amr-Shams Can you give us an example of creating the function that you are talking about ?

**Comment by Amr-Shams** [comment]
> > when it comes to this large size it is better to make it automatically by creating a function that takes the list and iterates over the edges and specified information and the function will do the job for your but still need to process the list before passing to the function.. like creating a list of objects...
> 
> @Amr-Shams Can you give us an example of creating the function that you are talking about ?
```SQL
CREATE OR REPLACE FUNCTION create_multiple_edges(nodes text[][]) RETURNS void AS $$
BEGIN
    FOREACH node IN ARRAY nodes
    LOOP
        EXECUTE format('MATCH (a:Person {name: %L}), (b:Location {name: %L}) CREATE (a)-[:VISITED]->(b)', node[1], node[2]);
    END LOOP;
END;
$$ LANGUAGE plpgsql;
```
as I mentioned you have to preprocess the text input so the mapping would be easy 
``` SQL
CREATE OR REPLACE FUNCTION merge_names_cities(names text[], cities text[]) RETURNS text[][] AS $$
DECLARE
    merged text[][];
BEGIN
    IF array_length(names, 1) <> array_length(cities, 1) THEN
        RAISE EXCEPTION 'Input arrays must be of equal length';
    END IF;

    FOR i IN 1..array_length(names, 1) LOOP
        merged[i][1] := names[i];
        merged[i][2] := cities[i];
    END LOOP;

    RETURN merged;
END;
$$ LANGUAGE plpgsql;
```
in the above function, the input should be easily merged and then the output of the above function should be passed to the edge_creation function .. assuming the input will be mapped in this way

**Comment by safi50** [comment]
Another way to do so is by using `UNWIND`.  Suppose I have 6 vertices: 

```
(:Person {name: 'Joe', title: 'Developer'}),  
(:Person {name: 'Andy', title: 'Manager'}),         
(:Person {name: 'Hakan', title: 'Protector'}),     
(:Person {name: 'Leyla', title: 'Assistant'}), 
(:Person {name: 'Sofi', title: 'Explorer'}),          
(:Person {name: 'Ian', title: 'Scientist'})   
```
Now, I can use `UWIND` to create edges for these with a single cypher query. See example below: 

```
SELECT * FROM cypher('test', $$
WITH [
  {source: 'Joe', target: 'Andy', weight: 2},
  {source: 'Andy', target: 'Hakan', weight: 3},
  {source: 'Andy', target: 'Leyla', weight: 1},
  {source: 'Hakan', target: 'Sofi', weight: 4},
  {source: 'Sofi', target: 'Ian', weight: 2}
] AS edges

UNWIND edges AS edge

MATCH (source:Person {name: edge.source})
MATCH (target:Person {name: edge.target})

CREATE (source)-[rel:WORKS_WITH {weight: edge.weight}]->(target) RETURN rel $$) as ( a agtype);
```

**Comment by hammadsaleemm** [comment]
Yes, you can use a single Cypher query to add multiple edges at a time in Neo4j graph database. The Cypher query language provides several ways to create multiple edges efficiently. My take is on UNWINDIND

One common way is to use the UNWIND clause to create multiple edges at once from a collection of edge properties. Here is an example: `UNWIND [{source: 1, target: 2}, {source: 2, target: 3}, {source: 3, target: 1}] AS edge
MATCH (s:Node {id: edge.source}), (t:Node {id: edge.target})
CREATE (s)-[:CONNECTS_TO]->(t)
`
In this example, the UNWIND clause unpacks a list of edge properties, each containing the source and target nodes' ids. The MATCH clause finds the corresponding nodes in the graph, and the CREATE clause creates the CONNECTS_TO relationship between them.

You can modify this query to add any other properties to the edges as needed. You can also adjust the MATCH clause to use different node labels or property names, depending on your graph's schema.

Note that adding multiple edges at once can still be slow if you have a large number of nodes or edges. In that case, you may want to consider using bulk import tools or using Neo4j's apoc library to optimize the import process.

**Comment by Munmud** [comment]
Thank you everyone
